src/controllers/libraryController.ts

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import * as fse from "fs-extra";
+import * as _ from "lodash";
+import { commands, Disposable, ExtensionContext, Uri, window, workspace, WorkspaceFolder } from "vscode";
+import { instrumentOperation } from "vscode-extension-telemetry-wrapper";
+import { Commands } from "../commands";
+import { Jdtls } from "../java/jdtls";
+import { Settings } from "../settings";
+import { Utility } from "../utility";
+import { DataNode } from "../views/dataNode";
+
+export class LibraryController implements Disposable {
+
+    private disposable: Disposable;
+
+    public constructor(public readonly context: ExtensionContext) {
+        this.disposable = Disposable.from(
+            commands.registerCommand(Commands.JAVA_PROJECT_ADD_LIBRARIES,
+                instrumentOperation(Commands.JAVA_PROJECT_ADD_LIBRARIES, (operationId: string, node: DataNode) => this.addLibraries())),
+            commands.registerCommand(Commands.JAVA_PROJECT_REMOVE_LIBRARY,
+                instrumentOperation(Commands.JAVA_PROJECT_REMOVE_LIBRARY, (operationId: string, node: DataNode) => this.removeLibrary(node.path))),
+            commands.registerCommand(Commands.JAVA_PROJECT_REFRESH_LIBRARIES,
+                instrumentOperation(Commands.JAVA_PROJECT_REFRESH_LIBRARIES, (operationId: string, node: DataNode) => this.refreshLibraries())),
+        );
+    }
+
+    public dispose() {
+        this.disposable.dispose();
+    }
+
+    public async addLibraries(libraryGlobs?: string[]) {
+        if (!libraryGlobs) {
+            libraryGlobs = [];
+            const workspaceFolder: WorkspaceFolder | undefined = Utility.getDefaultWorkspaceFolder();
+            const results: Uri[] | undefined = await window.showOpenDialog({
+                defaultUri: workspaceFolder && workspaceFolder.uri,
+                canSelectFiles: true,
+                canSelectFolders: true,
+                canSelectMany: true,
+                openLabel: "Select",
+            });
+            if (!results) {
+                return;
+            }
+            libraryGlobs = await Promise.all(results.map(async (uri: Uri) => {
+                const uriPath = workspace.asRelativePath(uri, false);
+                return (await fse.stat(uri.fsPath)).isDirectory() ? `${uriPath}/**/*.jar` : uriPath;
+            }));
+        }
+        const setting = Settings.referencedLibraries();
+        if (Settings.isDefaultReferencedLibraries()) {
```

Comment:
Why need check the default value?
why not use `setting.include = setting.include || []`?

Reply:
The `isDefaultReferencedLibraries` checks whether the `referencedLibraries` setting has been modified by user, instead of checking whether `setting.include` exists. In fact, `setting.include` always exists.

The motivation here is when a user adds a library at the first time, the default value `lib/**/*.jar` should not go with the added libraries. e.g., when adding `data/**/*.jar`, the setting in the workspace should be like this:
```json
"java.project.referencedLibraries": [
    "data/**/*.jar"
]
```
But without this check, it will be like this:
```json
"java.project.referencedLibraries": [
    "lib/**/*.jar",
    "data/**/*.jar"
]
```
Which will make the user confusing.

Reply:
But maybe this is also a confusing behavior. Because removing the "lib/**/*.jar" may cause some jar dependency lost. I will remove this line.